criterion-measurement/src/Criterion/Measurement.hs

```diff
@@ -289,41 +291,69 @@ runBenchmark :: Benchmarkable
              -- exceeded in order to generate enough data to perform
              -- meaningful statistical analyses.
              -> IO (V.Vector Measured, Double)
-runBenchmark bm timeLimit = do
+runBenchmark bm timeLimit = runBenchmarkWith endAt 0 bm where
+  endAt :: Int -> Int64 -> Double -> NonEmpty Measured -> Double -> Maybe (Int64, Double)
+  endAt count iters delta ms prev = do
+    let m = NE.head ms
+
+    -- we accumulate the time spent in runs longer then threshold.
+    let prev' | measTime m >= threshold = prev + measTime m
+              | otherwise               = prev
+
+    -- We try to honour the time limit, but we also have more
+    -- important constraints:
+    --
+    -- We must generate enough data that bootstrapping won't
+    -- simply crash.
+    --
+    -- We need to generate enough measurements that have long
+    -- spans of execution to outweigh the (rather high) cost of
+    -- measurement.
+
+    if delta >= timeLimit && prev' >= threshold * 10 && count >= 4
+    then Nothing
+    else Just (nextIters iters, prev')
+
+  -- Multiply by 1.05, but always increase
+  nextIters :: Int64 -> Int64
+  nextIters n = max (n + 1) (div (105 * n) 100)
+
+-- | Run benchmark with provided dynamic termination condition.
+--
+-- The first argument is termination check, it takes as arguments:
+--
+-- * current run count (starting from zero)
```

Comment:
If `TerminationCheck` were factored out into a newtype, these Haddocks could be moved to the newtype instead.